src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                }
+
+                return _currentBasicBlock;
+            }
+        }
+
+        private void AddStatement(IOperation statement)
+        {
+            if (statement == null)
+            {
+                return;
+            }
+
+            // PROTOTYPE(dataflow): Assert that statement's parent is null at this point.
+            CurrentBasicBlock.AddStatement(statement);
+        }
+
+        private void AppendNewBlock(BasicBlock block)
+        {
+            Debug.Assert(block != null);
+            BasicBlock prevBlock = _blocks.Last();
+
+            if (prevBlock.Next == null)
+            {
+                LinkBlocks(prevBlock, block);
+            }
+
+            _blocks.Add(block);
+            _currentBasicBlock = block;
+        }
+
+        private static void LinkBlocks(BasicBlock prevBlock, BasicBlock nextBlock)
+        {
+            Debug.Assert(prevBlock.Next == null);
+            prevBlock.Next = nextBlock;
+            nextBlock.AddPredecessor(prevBlock);
+        }
+
+        public override IOperation VisitBlock(IBlockOperation operation, object argument)
+        {
+            foreach (var statement in operation.Operations)
+            {
+                VisitStatement(statement);
+            }
+
+            return null;
+        }
+
+        public override IOperation VisitConditional(IConditionalOperation operation, object argument)
+        {
+            if (operation == _currentStatement)
+            {
+                if (operation.WhenFalse == null)
+                {
+                    // if (condition) 
+                    //   consequence;  
+                    //
+                    // becomes
+                    //
+                    // GotoIfFalse condition afterif;
+                    // consequence;
+                    // afterif:
+
+                    BasicBlock afterIf = null;
+                    VisitConditionalBranch(operation.Condition, ref afterIf, false);
+                    VisitStatement(operation.WhenTrue);
+                    AppendNewBlock(afterIf);
+                }
+                else
+                {
+                    // if (condition)
+                    //     consequence;
+                    // else 
+                    //     alternative
+                    //
+                    // becomes
+                    //
+                    // GotoIfFalse condition alt;
+                    // consequence
+                    // goto afterif;
+                    // alt:
+                    // alternative;
+                    // afterif:
+
+                    BasicBlock whenFalse = null;
+                    VisitConditionalBranch(operation.Condition, ref whenFalse, false);
+
+                    VisitStatement(operation.WhenTrue);
+
+                    var afterIf = new BasicBlock(BasicBlockKind.Block);
+                    LinkBlocks(CurrentBasicBlock, afterIf);
+                    _currentBasicBlock = null;
```

Comment:
>_currentBasicBlock [](start = 20, length = 18)

I'm not clear what is the purpose of setting `_currentBasicBlock` to `null`, since `AppendNewBlock` doesn't read it, but always sets it.

Reply:
> I'm not clear what is the purpose of setting _currentBasicBlock to null, since AppendNewBlock doesn't read it, but always sets it.

The purpose is to explicitly clear the current block to avoid adding any more statements to it by accident.

---
In reply to: [162152969](https://github.com/dotnet/roslyn/pull/24263#discussion_r162152969) [](ancestors = 162152969)